sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java

```diff
@@ -0,0 +1,117 @@
+import org.sonar.api.batch.sensor.highlighting.TypeOfText;
+import org.sonar.flex.api.FlexKeyword;
+import org.sonar.flex.api.FlexPunctuator;
+import org.sonar.flex.api.FlexTokenType;
+import org.sonar.squidbridge.SquidAstVisitor;
+import org.sonar.sslr.parser.LexerlessGrammar;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Set;
+
+public class FlexTokensVisitor extends SquidAstVisitor<LexerlessGrammar> {
+
+  private static final String NORMALIZED_CHARACTER_LITERAL = "$CHARS";
+  private static final String NORMALIZED_NUMERIC_LITERAL = "$NUMBER";
+  private static final Set<FlexKeyword> KEYWORDS = Sets.immutableEnumSet(Arrays.asList(FlexKeyword.values()));
+
+  private final SensorContext context;
+  private final Lexer lexer;
+
+  public FlexTokensVisitor(SensorContext context, Lexer lexer) {
+    this.context = context;
+    this.lexer = lexer;
+  }
+
+  @Override
+  public void visitFile(@Nullable AstNode astNode) {
+    NewHighlighting highlighting = context.newHighlighting();
+    File file = getContext().getFile();
+    InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates().is(file));
+    highlighting.onFile(inputFile);
+
+    NewCpdTokens cpdTokens = context.newCpdTokens();
+    cpdTokens.onFile(inputFile);
+
+    Iterator<Token> iterator = lexer.lex(file).iterator();
+    // we currently use this hack to remove "import" directives
+    boolean importDirective = false;
+    while (iterator.hasNext()) {
+      Token token = iterator.next();
+      TokenType tokenType = token.getType();
+      if (tokenType == FlexKeyword.IMPORT) {
+        importDirective = true;
+      } else if (importDirective) {
+        // We do nothing as we want to ignore "import" directives
+        if (tokenType == FlexPunctuator.SEMI) {
+          importDirective = false;
+        }
+      } else if (tokenType != GenericTokenType.EOF) {
+        TokenLocation tokenLocation = new TokenLocation(token);
+        cpdTokens.addToken(tokenLocation.startLine(), tokenLocation.startCharacter(), tokenLocation.endLine(), tokenLocation.endCharacter(), getTokenImage(token));
+      }
+      if (tokenType.equals(FlexTokenType.NUMERIC_LITERAL)) {
+        highlight(highlighting, token, TypeOfText.CONSTANT);
+      } else if (tokenType.equals(GenericTokenType.LITERAL)) {
+        highlight(highlighting, token, TypeOfText.STRING);
+      } else if (KEYWORDS.contains(tokenType)) {
+        highlight(highlighting, token, TypeOfText.KEYWORD);
+      }
+      for (Trivia trivia : token.getTrivia()) {
+        highlight(highlighting, trivia.getToken(), TypeOfText.COMMENT);
+      }
+    }
+
+    highlighting.save();
+    cpdTokens.save();
+  }
+
+  private static String getTokenImage(Token token) {
+    if (token.getType() == GenericTokenType.LITERAL) {
+      return NORMALIZED_CHARACTER_LITERAL;
+    } else if (token.getType() == FlexTokenType.NUMERIC_LITERAL) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)